### DISCUSSION TITLE
Component Designpattern

### DISCUSSION THREAD

**Discussion by AMatutat** [Programmiermethoden/Dungeon#81] [Ideas]
**Was ist das Entity-Component-System (ECS)?**

Entity: Ist ein leerer Container, der ein oder mehrere Components haben kann.
Components: Beschreiben die Verhalten/Eigenschaften einer Entität (hat Physik, hat Inventar, hat Textur, etc.) 
System: agiert auf Components (z. B. Physik-System nimmt Einfluss auf Physic-Components) und updatet die Werte

Mit einem ECS würden wir daher weg vom starken OOP Ansatz gehen. Weniger IST ein und mehr HAT ein. 

**Welche Vorteile hat es, zum jetzigen Vorgehen**

- deutlich bessere Code-Struktur, da wir die Vererbungshierarchie entschlacken + weniger duplizierter Code
- Entitäten könnten sich mit den Components einfach "zusammenstecken" lassen
- Vereinfacht die DSL Eingabe + erlaubt mehr Möglichkeiten, da die Components beliebig kombiniert werden können

**ECS in unserem Dungeon**
- Nicht in das Framework integrieren, [alte Diskussion zu dem Thema](https://github.com/Programmiermethoden/PM-Dungeon/issues/323)
- viele unserer Interfaces (z. B. `IAnimatable`) würden zu Components werden
- Ausklemmen der Dungeon-Controller und dafür die Systeme einklinken
- Wir geben Components und Systems vor, DSL steckt zusammen
- Components sollten Default verhalten haben
- EventHandler in den components (?), um eigenes Verhalten in der DSL zu definieren , Listener-Pattern "gibt es was in der DSL, sonst Standard verhalten

**Comment by malt-r** [comment]
> Warum ist das besser als das jetzige System?
- Über DSL sind eigene Objekte einfacher zu erstellen, indem Komponenten kombiniert werden

**Comment by malt-r** [comment]
https://gameprogrammingpatterns.com/component.html
